nixos/modules/services/networking/wireguard.nix

```diff
@@ -153,35 +182,43 @@ let
   wgCommand = "${pkgs.wireguard}/bin/wg";
 
   generateUnit = name: values:
+    assert (values.privateKey != null || values.privateKeyFile != null);
     nameValuePair "wireguard-${name}"
       {
         description = "WireGuard Tunnel - ${name}";
         after = [ "network.target" ];
         wantedBy = [ "multi-user.target" ];
+        script = builtins.concatStringsSep "\n" (lib.flatten([
+          values.preSetup
+
+          ''
+          ${ipCommand} link del dev "${name}" || true
+          ${ipCommand} link add dev ${name} type wireguard
+          ${wgCommand} setconf ${name} /dev/stdin <<EOF
```

Comment:
I would use `wg set` instead of creating a wireguard configuration file. Then a key file can be used directly without the need for here-doc. The old attributes can be still used in this scheme using command substitution `<(echo $key)`